### PR TITLE
Fix random e2e failure.

### PR DESCRIPTION
**Pull request by pratikjagrut**
### Motivation
fix E2E failure

### Changes
- Seed random with time.Now to generate unique digits on every subtest call.

- Replace retry code block with require.Eventually.

- Remove test timeout from makefile test-e2e target.

- Update github.com/stretchr/testify from v1.4.0 to v1.6.1
	The 1.4.0 version we were using had a bug in the Eventually function,
	which caused flaky results.

- Increasing timeout for Eventually function

- Update vendor dependencies.

- Updating test timeout.

### Testing

`make test-e2e`

### CODE REVIEW

**Comment by DhritiShikhar** [comment]
Tested this PR locally. E2E passes for me locally on 4.4 cluster.

File: Makefile [hunk 1]
```diff
@@ -220,7 +220,7 @@ test-e2e: e2e-setup deploy-crds
 		operator-sdk --verbose test local ./test/e2e \
 			--namespace $(TEST_NAMESPACE) \
 			--up-local \
-			--go-test-flags "-timeout=15m" \
+			--go-test-flags "-timeout=108m" \
```
Comment (Avni-Sharma) [inline Makefile:223]: Hi @pratikjagrut any specific reason to bump the timeout to 1.8 hrs from 15 mins?

Comment (pratikjagrut) [reply Makefile:223]: I've mentioned it in my commit message.

Comment (pratikjagrut) [reply Makefile:223]: Worst case timeout is 108min, calculation as follows:

6 eventually loops each timeout 2min: 6 * 2 = 12min
9 sequential subtests, timeout: 9 * 12 = 108min.

**Comment by Avni-Sharma** [comment]
tried `make-e2e` on cluster 4.5. Encountered the following errors

```
  TestAddSchemesToFramework/end-to-end/scenario-db-app-sbr: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-db-app-sbr: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-db-app-sbr: servicebindingrequest_test.go:551: 
        	Error Trace:	servicebindingrequest_test.go:551
        	            				servicebindingrequest_test.go:155
        	            				servicebindingrequest_test.go:91
        	Error:      	Condition never satisfied
```

```
 TestAddSchemesToFramework/end-to-end/scenario-app-db-sbr: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-app-db-sbr: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-app-db-sbr: servicebindingrequest_test.go:551: 
        	Error Trace:	servicebindingrequest_test.go:551
        	            				servicebindingrequest_test.go:155
        	            				servicebindingrequest_test.go:94
        	Error:      	Condition never satisfied
        	Test:       	TestAddSchemesToFramework/end-to-end/scenario-app-db-sbr
time="2020-07-07T18:52:54+05:30" level=info msg="Skipping cleanup function since --skip-cleanup-error is true"
=== RUN   TestAddSchemesToFramework/end-to-end/scenario-db-sbr-app
```

```
 TestAddSchemesToFramework/end-to-end/scenario-db-sbr-app: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-db-sbr-app: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-db-sbr-app: servicebindingrequest_test.go:551: 
        	Error Trace:	servicebindingrequest_test.go:551
        	            				servicebindingrequest_test.go:155
        	            				servicebindingrequest_test.go:97
        	Error:      	Condition never satisfied
        	Test:       	TestAddSchemesToFramework/end-to-end/scenario-db-sbr-app
time="2020-07-07T18:55:17+05:30" level=info msg="Skipping cleanup function since --skip-cleanup-error is true"
=== RUN   TestAddSchemesToFramework/end-to-end/scenario-app-sbr-db
```

```
TestAddSchemesToFramework/end-to-end/scenario-app-sbr-db: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-app-sbr-db: servicebindingrequest_test.go:551: 
        	Error Trace:	servicebindingrequest_test.go:551
        	            				servicebindingrequest_test.go:155
        	            				servicebindingrequest_test.go:100
        	Error:      	Condition never satisfied
        	Test:       	TestAddSchemesToFramework/end-to-end/scenario-app-sbr-db
time="2020-07-07T18:57:35+05:30" level=info msg="Skipping cleanup function since --skip-cleanup-error is true"
```

```
    TestAddSchemesToFramework/end-to-end/scenario-sbr-db-app: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-sbr-db-app: servicebindingrequest_test.go:551: 
        	Error Trace:	servicebindingrequest_test.go:551
        	            				servicebindingrequest_test.go:155
        	            				servicebindingrequest_test.go:103
        	Error:      	Condition never satisfied
        	Test:       	TestAddSchemesToFramework/end-to-end/scenario-sbr-db-app
time="2020-07-07T19:00:07+05:30" level=info msg="Skipping cleanup function since --skip-cleanup-error is true"
=== RUN   TestAddSchemesToFramework/end-to-end/scenario-sbr-app-db
```

```

    TestAddSchemesToFramework/end-to-end/scenario-sbr-app-db: servicebindingrequest_test.go:558: binding secret contents are invalid: nil
    TestAddSchemesToFramework/end-to-end/scenario-sbr-app-db: servicebindingrequest_test.go:551: 
        	Error Trace:	servicebindingrequest_test.go:551
        	            				servicebindingrequest_test.go:155
        	            				servicebindingrequest_test.go:106
        	Error:      	Condition never satisfied
        	Test:       	TestAddSchemesToFramework/end-to-end/scenario-sbr-app-db
time="2020-07-07T19:02:25+05:30" level=info msg="Skipping cleanup function since --skip-cleanup-error is true"
```

```
    TestAddSchemesToFramework/end-to-end/scenario-csv-db-app-sbr: servicebindingrequest_test.go:551: 
        	Error Trace:	servicebindingrequest_test.go:551
        	            				servicebindingrequest_test.go:155
        	            				servicebindingrequest_test.go:109
        	Error:      	Condition never satisfied
        	Test:       	TestAddSchemesToFramework/end-to-end/scenario-csv-db-app-sbr
time="2020-07-07T19:04:44+05:30" level=info msg="Skipping cleanup function since --skip-cleanup-error is true"
```

**Comment by pratikjagrut** [comment]
@Avni-Sharma I'm not sure what exactly happens on a local run what is producing this error but on CI this seems not to be a case.

**Comment by pratikjagrut** [comment]
@Avni-Sharma I just tested e2e locally on 4.5 cluster they passed.
```
--- PASS: TestAddSchemesToFramework (573.91s)
    --- PASS: TestAddSchemesToFramework/end-to-end (401.57s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-etcd-unannotated-app-db-sbr (59.81s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-db-app-sbr (47.31s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-app-db-sbr (42.63s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-db-sbr-app (37.76s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-app-sbr-db (42.23s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-sbr-db-app (40.32s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-sbr-app-db (41.28s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-csv-db-app-sbr (40.96s)
        --- PASS: TestAddSchemesToFramework/end-to-end/scenario-csv-app-db-sbr (49.28s)
PASS
```

File: Makefile (hunk 1 above)
Comment (pratikjagrut) [reply Makefile:223]: @Avni-Sharma Should I resolve this conversation?

**Comment by Avni-Sharma** [comment]
> @Avni-Sharma I'm not sure what exactly happens on a local run to produce this error. Looks like secrets are empty, but on CI this error has not occurred yet.

Thanks @pratikjagrut . I will try running `make test-e2e`  locally again for cluster 4.5 .

**Comment by Avni-Sharma** [comment]
Passes on a fresh cluster for me now.

File: vendor/modules.txt [hunk 2]
```diff
@@ -192,7 +192,7 @@ github.com/spf13/cobra
 # github.com/spf13/pflag v1.0.5
 ## explicit
 github.com/spf13/pflag
-# github.com/stretchr/testify v1.4.0
+# github.com/stretchr/testify v1.6.1
```
Comment (akashshinde) [inline vendor/modules.txt:195]: @pratikjagrut any specific reason we're upgrading package in this PR ?

Comment (akashshinde) [reply vendor/modules.txt:195]: got it

Comment (pratikjagrut) [reply vendor/modules.txt:195]: Resolving this conversation.

File: test/e2e/servicebindingrequest_test.go [hunk 3]
```diff
@@ -583,7 +584,7 @@ func serviceBindingRequestTest(
 			return false
 		}
 		return true
-	}, 60*time.Second, 2*time.Second)
+	}, 600*time.Second, 2*time.Second)
```
[the remark this answers is not in the document]
Comment (pratikjagrut) [reply test/e2e/servicebindingrequest_test.go]: This is an outdated code, I increased the timeout as it was running into `condition not satisfied` error which usually comes after Eventually loop timeouts.

**Comment by isutton** [comment]
> @pratikjagrut any specific reason we're upgrading package in this PR ? 

From the PR's description:

```
Update github.com/stretchr/testify from v1.4.0 to v1.6.1
The 1.4.0 version we were using had a bug in the Eventually function,
which caused flaky results.
```

**Comment by pratikjagrut** [comment]
I pushed squashed commit.

**Comment by pmacik** [comment]
Looks like the e2e tests failed for the infinite loop (https://github.com/redhat-developer/service-binding-operator/pull/523)

**Comment by pratikjagrut** [comment]
> Looks like the e2e tests failed for the infinite loop (#523)

@pmacik One reason I could think of is if we create SBR before the application and then the operator is watching for the application deployment to be ready, and at the same time `e2eutil.WaitForDeployment(t, f.KubeClient, ns, appName, 1, retryInterval, timeout)` is also watching. So here this two watches are in deadlock or sort of deadlock if operator watch detects the deployment is ready before e2eutil watch then it will redeploy it with the secrets hence the e2eutil watch might miss the first deployment readiness and will still wait for deployment this time deployment will take some more time to reach 1 replica and e2eutil watch will timeout. Hence failing the test.

**Comment by Avni-Sharma** [comment]
Hi @pratikjagrut can you make an issue for the above-mentioned scenario so that we can look it up again.

**Pull request merged**